### ISSUE TITLE
tsc: symlink: node_modules->../node_modules: results in selenium "Duplicate identifier"

### ISSUE THREAD

**Issue by teckno-query** [Microsoft/TypeScript#12619]
**Issue**
In the TypeScript "quickstart" application directory, when "node_modules" is a symbolic link to "../node_modules", during "tsc" compilation, the global variables in selenium-webdriver cause "Duplicate identifier" errors.  This appears to be due to the default behavior of tsconfig.

@see
https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#types-typeroots-and-types

If the following "typesRoots" attribute is added to "tsconfig.json", the errors go away:
```
{
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs",
    "moduleResolution": "node",
    "sourceMap": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": [ "es2015", "dom" ],
    "noImplicitAny": true,
    "suppressImplicitAnyIndexErrors": true,
    "typeRoots": ["../node_modules/@types"]
  },
  "exclude": [
    "node_modules",
    "**/*-aot.ts"
  ]
}
```

However, this workaround is undesired since the project should be able to compile correctly whether the "./node_modules" is an actual directory or a symbolic link.  And, just to be clear, when "./node_modules" is an actual directory, the compilation of the TypeScript "quickstart" application is successful out-of-the-box.  So, this seems to be a typescript "tsc" issue with symbolic links.

**Steps to Reproduce**
1. Download quickstart project from https://angular.io/docs/ts/latest/quickstart.html
`git clone https://github.com/angular/quickstart.git quickstart`
2. In the directory containing quickstart, create directory "node_modules"
```
$ mkdir node_modules
$ ls -l
drwxr-xr-x  364 user  wheel  12376 Dec  2 11:02 node_modules
drwxr-xr-x   24 user  wheel    816 Nov 24 12:37 quickstart
```
3. Change into "quickstart" directory and create link to ../node_modules
```
$ cd quickstart
$ ln -s ../node_modules .
$ ls -l
...
lrwxr-xr-x   1 user  wheel    15 Nov 24 12:37 node_modules -> ../node_modules
...
```
4. Install packages (note: these packages will be installed in ../node_modules)
`$ npm install`
5. Compile
`$ tsc`
6. Issue: "Duplicate identifier ..."; for example:
```
node_modules/@types/selenium-webdriver/index.d.ts(12,11): error TS2300: Duplicate identifier 'Driver'.

../node_modules/@types/selenium-webdriver/index.d.ts(12,11): error TS2300: Duplicate identifier 'Driver'.
```

**Versions**
```
$ node -v
v7.2.0
$ npm -v
4.0.3
$ tsc -v
Version 2.0.10
```

**Code**

```
$  git clone https://github.com/angular/quickstart.git quickstart
Cloning into 'quickstart'...
remote: Counting objects: 1237, done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 1237 (delta 4), reused 0 (delta 0), pack-reused 1220
Receiving objects: 100% (1237/1237), 1.09 MiB | 707.00 KiB/s, done.
Resolving deltas: 100% (675/675), done.
Checking connectivity... done.
$
$  mkdir node_modules
$
$  cd quickstart/
$
$  ln -s ../node_modules .
$
$  ls -l
total 152
-rw-r--r--  1 user  wheel  4253 Dec  2 12:53 CHANGELOG.md
-rw-r--r--  1 user  wheel  1078 Dec  2 12:53 LICENSE
-rw-r--r--  1 user  wheel  5726 Dec  2 12:53 README.md
drwxr-xr-x  6 user  wheel   204 Dec  2 12:53 app
drwxr-xr-x  3 user  wheel   102 Dec  2 12:53 e2e
-rw-r--r--  1 user  wheel  5430 Dec  2 12:53 favicon.ico
-rw-r--r--  1 user  wheel   774 Dec  2 12:53 index.html
-rw-r--r--  1 user  wheel  3238 Dec  2 12:53 karma-test-shim.js
-rw-r--r--  1 user  wheel  3589 Dec  2 12:53 karma.conf.js
lrwxr-xr-x  1 user  wheel    15 Dec  2 12:54 node_modules -> ../node_modules
-rw-r--r--  1 user  wheel  1849 Dec  2 12:53 package.json
-rw-r--r--  1 user  wheel  5049 Dec  2 12:53 protractor.config.js
-rw-r--r--  1 user  wheel    91 Dec  2 12:53 styles.css
-rw-r--r--  1 user  wheel   209 Dec  2 12:53 systemjs.config.extras.js
-rw-r--r--  1 user  wheel  1454 Dec  2 12:53 systemjs.config.js
-rw-r--r--  1 user  wheel   306 Dec  2 12:53 tsconfig.json
-rw-r--r--  1 user  wheel  1896 Dec  2 12:53 tslint.json
$
$  npm install
npm WARN deprecated node-uuid@1.4.7: use uuid module instead
npm WARN deprecated tough-cookie@2.2.2: ReDoS vulnerability parsing Set-Cookie https://nodesecurity.io/advisories/130

> fsevents@1.0.15 install /Users/Shared/working/WebstormProjects/test/quickstart/node_modules/fsevents
> node-pre-gyp install --fallback-to-build

[fsevents] Success: "/Users/Shared/working/WebstormProjects/test/node_modules/fsevents/lib/binding/Release/node-v51-darwin-x64/fse.node" is installed via remote
angular-quickstart@1.0.0 /Users/Shared/working/WebstormProjects/test/quickstart
...
$
$  tsc
```

**Expected behavior**
```
$ tsc
<no errors>
```

**Actual behavior**
```
$ tsc
../node_modules/@types/selenium-webdriver/index.d.ts(12,11): error TS2300: Duplicate identifier 'Driver'.
../node_modules/@types/selenium-webdriver/index.d.ts(46,11): error TS2300: Duplicate identifier 'Options'.
../node_modules/@types/selenium-webdriver/index.d.ts(310,11): error TS2300: Duplicate identifier 'ServiceBuilder'.
../node_modules/@types/selenium-webdriver/index.d.ts(422,11): error TS2300: Duplicate identifier 'Driver'.
../node_modules/@types/selenium-webdriver/index.d.ts(444,11): error TS2300: Duplicate identifier 'Options'.
../node_modules/@types/selenium-webdriver/index.d.ts(484,11): error TS2300: Duplicate identifier 'ServiceBuilder'.
../node_modules/@types/selenium-webdriver/index.d.ts(561,9): error TS2300: Duplicate identifier 'Binary'.
../node_modules/@types/selenium-webdriver/index.d.ts(599,9): error TS2300: Duplicate identifier 'Profile'.
../node_modules/@types/selenium-webdriver/index.d.ts(718,9): error TS2300: Duplicate identifier 'Options'.
../node_modules/@types/selenium-webdriver/index.d.ts(790,9): error TS2300: Duplicate identifier 'Driver'.
../node_modules/@types/selenium-webdriver/index.d.ts(824,9): error TS2300: Duplicate identifier 'HttpRequest'.
../node_modules/@types/selenium-webdriver/index.d.ts(840,9): error TS2300: Duplicate identifier 'HttpResponse'.
../node_modules/@types/selenium-webdriver/index.d.ts(862,9): error TS2300: Duplicate identifier 'HttpClient'.
../node_modules/@types/selenium-webdriver/index.d.ts(908,9): error TS2300: Duplicate identifier 'Executor'.
../node_modules/@types/selenium-webdriver/index.d.ts(968,9): error TS2300: Duplicate identifier 'Driver'.
../node_modules/@types/selenium-webdriver/index.d.ts(988,9): error TS2300: Duplicate identifier 'Options'.
../node_modules/@types/selenium-webdriver/index.d.ts(1177,9): error TS2300: Duplicate identifier 'ServiceBuilder'.
../node_modules/@types/selenium-webdriver/index.d.ts(1260,9): error TS2300: Duplicate identifier 'Options'.
../node_modules/@types/selenium-webdriver/index.d.ts(1324,9): error TS2300: Duplicate identifier 'Driver'.
../node_modules/@types/selenium-webdriver/index.d.ts(1361,9): error TS2300: Duplicate identifier 'DriverService'.
../node_modules/@types/selenium-webdriver/index.d.ts(1413,9): error TS2300: Duplicate identifier 'Server'.
../node_modules/@types/selenium-webdriver/index.d.ts(1441,9): error TS2300: Duplicate identifier 'CommandExecutor'.
../node_modules/@types/selenium-webdriver/index.d.ts(1447,9): error TS2300: Duplicate identifier 'Options'.
../node_modules/@types/selenium-webdriver/index.d.ts(1490,9): error TS2300: Duplicate identifier 'Driver'.
../node_modules/@types/selenium-webdriver/index.d.ts(1505,15): error TS2300: Duplicate identifier 'IError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1515,15): error TS2300: Duplicate identifier 'WebDriverError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1523,15): error TS2300: Duplicate identifier 'ElementNotSelectableError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1532,15): error TS2300: Duplicate identifier 'ElementNotVisibleError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1540,15): error TS2300: Duplicate identifier 'InvalidArgumentError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1549,15): error TS2300: Duplicate identifier 'InvalidCookieDomainError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1557,15): error TS2300: Duplicate identifier 'InvalidElementCoordinatesError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1567,15): error TS2300: Duplicate identifier 'InvalidElementStateError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1575,15): error TS2300: Duplicate identifier 'InvalidSelectorError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1583,15): error TS2300: Duplicate identifier 'NoSuchSessionError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1591,15): error TS2300: Duplicate identifier 'JavascriptError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1600,15): error TS2300: Duplicate identifier 'MoveTargetOutOfBoundsError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1608,15): error TS2300: Duplicate identifier 'NoSuchAlertError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1617,15): error TS2300: Duplicate identifier 'NoSuchElementError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1626,15): error TS2300: Duplicate identifier 'NoSuchFrameError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1635,15): error TS2300: Duplicate identifier 'NoSuchWindowError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1643,15): error TS2300: Duplicate identifier 'ScriptTimeoutError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1651,15): error TS2300: Duplicate identifier 'SessionNotCreatedError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1660,15): error TS2300: Duplicate identifier 'StaleElementReferenceError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1668,15): error TS2300: Duplicate identifier 'TimeoutError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1676,15): error TS2300: Duplicate identifier 'UnableToSetCookieError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1684,15): error TS2300: Duplicate identifier 'UnableToCaptureScreenError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1692,15): error TS2300: Duplicate identifier 'UnexpectedAlertOpenError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1709,15): error TS2300: Duplicate identifier 'UnknownCommandError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1718,15): error TS2300: Duplicate identifier 'UnknownMethodError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1726,15): error TS2300: Duplicate identifier 'UnsupportedOperationError'.
../node_modules/@types/selenium-webdriver/index.d.ts(1738,15): error TS2300: Duplicate identifier 'Preferences'.
../node_modules/@types/selenium-webdriver/index.d.ts(1767,15): error TS2300: Duplicate identifier 'Level'.
../node_modules/@types/selenium-webdriver/index.d.ts(1852,15): error TS2300: Duplicate identifier 'Entry'.
../node_modules/@types/selenium-webdriver/index.d.ts(1899,15): error TS2300: Duplicate identifier 'Logger'.
../node_modules/@types/selenium-webdriver/index.d.ts(2023,15): error TS2300: Duplicate identifier 'LogManager'.
../node_modules/@types/selenium-webdriver/index.d.ts(2288,15): error TS2300: Duplicate identifier 'CancellationError'.
../node_modules/@types/selenium-webdriver/index.d.ts(2357,15): error TS2300: Duplicate identifier 'Thenable'.
../node_modules/@types/selenium-webdriver/index.d.ts(2488,15): error TS2300: Duplicate identifier 'Promise'.
../node_modules/@types/selenium-webdriver/index.d.ts(2643,15): error TS2300: Duplicate identifier 'Deferred'.
../node_modules/@types/selenium-webdriver/index.d.ts(2758,15): error TS2300: Duplicate identifier 'ControlFlow'.
../node_modules/@types/selenium-webdriver/index.d.ts(2866,15): error TS2300: Duplicate identifier 'Condition'.
../node_modules/@types/selenium-webdriver/index.d.ts(3177,11): error TS2300: Duplicate identifier 'ActionSequence'.
../node_modules/@types/selenium-webdriver/index.d.ts(3365,11): error TS2300: Duplicate identifier 'TouchSequence'.
../node_modules/@types/selenium-webdriver/index.d.ts(3489,11): error TS2300: Duplicate identifier 'Alert'.
../node_modules/@types/selenium-webdriver/index.d.ts(3561,11): error TS2300: Duplicate identifier 'AlertPromise'.
../node_modules/@types/selenium-webdriver/index.d.ts(3694,11): error TS2300: Duplicate identifier 'UnhandledAlertError'.
../node_modules/@types/selenium-webdriver/index.d.ts(3766,11): error TS2300: Duplicate identifier 'Builder'.
../node_modules/@types/selenium-webdriver/index.d.ts(4011,11): error TS2300: Duplicate identifier 'By'.
../node_modules/@types/selenium-webdriver/index.d.ts(4135,10): error TS2300: Duplicate identifier 'ByHash'.
../node_modules/@types/selenium-webdriver/index.d.ts(4241,11): error TS2300: Duplicate identifier 'Capabilities'.
../node_modules/@types/selenium-webdriver/index.d.ts(4534,11): error TS2300: Duplicate identifier 'Command'.
../node_modules/@types/selenium-webdriver/index.d.ts(4586,11): error TS2300: Duplicate identifier 'Executor'.
../node_modules/@types/selenium-webdriver/index.d.ts(4605,11): error TS2300: Duplicate identifier 'DeferredExecutor'.
../node_modules/@types/selenium-webdriver/index.d.ts(4616,11): error TS2300: Duplicate identifier 'Listener'.
../node_modules/@types/selenium-webdriver/index.d.ts(4631,11): error TS2300: Duplicate identifier 'EventEmitter'.
../node_modules/@types/selenium-webdriver/index.d.ts(4713,11): error TS2300: Duplicate identifier 'Navigation'.
../node_modules/@types/selenium-webdriver/index.d.ts(4776,11): error TS2300: Duplicate identifier 'Options'.
../node_modules/@types/selenium-webdriver/index.d.ts(4866,11): error TS2300: Duplicate identifier 'Timeouts'.
../node_modules/@types/selenium-webdriver/index.d.ts(4928,11): error TS2300: Duplicate identifier 'Window'.
../node_modules/@types/selenium-webdriver/index.d.ts(4991,11): error TS2300: Duplicate identifier 'Logs'.
../node_modules/@types/selenium-webdriver/index.d.ts(5034,11): error TS2300: Duplicate identifier 'TargetLocator'.
../node_modules/@types/selenium-webdriver/index.d.ts(5126,11): error TS2300: Duplicate identifier 'FileDetector'.
../node_modules/@types/selenium-webdriver/index.d.ts(5167,11): error TS2300: Duplicate identifier 'WebDriver'.
../node_modules/@types/selenium-webdriver/index.d.ts(5634,9): error TS2374: Duplicate string index signature.
../node_modules/@types/selenium-webdriver/index.d.ts(5962,11): error TS2300: Duplicate identifier 'WebElement'.
../node_modules/@types/selenium-webdriver/index.d.ts(6319,11): error TS2300: Duplicate identifier 'WebElementPromise'.
../node_modules/@types/selenium-webdriver/index.d.ts(6467,11): error TS2300: Duplicate identifier 'Session'.
../node_modules/@types/selenium-webdriver/index.d.ts(6574,5): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6578,3): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6582,5): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6586,3): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6590,3): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6594,3): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6598,3): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6602,3): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6606,3): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6610,5): error TS2300: Duplicate identifier 'export='.
../node_modules/@types/selenium-webdriver/index.d.ts(6614,5): error TS2300: Duplicate identifier 'export='.
node_modules/@types/selenium-webdriver/index.d.ts(12,11): error TS2300: Duplicate identifier 'Driver'.
node_modules/@types/selenium-webdriver/index.d.ts(46,11): error TS2300: Duplicate identifier 'Options'.
node_modules/@types/selenium-webdriver/index.d.ts(310,11): error TS2300: Duplicate identifier 'ServiceBuilder'.
node_modules/@types/selenium-webdriver/index.d.ts(422,11): error TS2300: Duplicate identifier 'Driver'.
node_modules/@types/selenium-webdriver/index.d.ts(444,11): error TS2300: Duplicate identifier 'Options'.
node_modules/@types/selenium-webdriver/index.d.ts(484,11): error TS2300: Duplicate identifier 'ServiceBuilder'.
node_modules/@types/selenium-webdriver/index.d.ts(561,9): error TS2300: Duplicate identifier 'Binary'.
node_modules/@types/selenium-webdriver/index.d.ts(599,9): error TS2300: Duplicate identifier 'Profile'.
node_modules/@types/selenium-webdriver/index.d.ts(718,9): error TS2300: Duplicate identifier 'Options'.
node_modules/@types/selenium-webdriver/index.d.ts(790,9): error TS2300: Duplicate identifier 'Driver'.
node_modules/@types/selenium-webdriver/index.d.ts(824,9): error TS2300: Duplicate identifier 'HttpRequest'.
node_modules/@types/selenium-webdriver/index.d.ts(840,9): error TS2300: Duplicate identifier 'HttpResponse'.
node_modules/@types/selenium-webdriver/index.d.ts(862,9): error TS2300: Duplicate identifier 'HttpClient'.
node_modules/@types/selenium-webdriver/index.d.ts(908,9): error TS2300: Duplicate identifier 'Executor'.
node_modules/@types/selenium-webdriver/index.d.ts(968,9): error TS2300: Duplicate identifier 'Driver'.
node_modules/@types/selenium-webdriver/index.d.ts(988,9): error TS2300: Duplicate identifier 'Options'.
```

**Comment by kevinvergauwen** [comment]
I have a related issue. If a file is a symlink the compiler uses the attributes of the target file, including the filename and path of the target file. So any path is relative to that file.

**Comment by mhegazy** [comment]
the core of the issue here is that the declaration of selenium-webdriver is not correct. it is authored as a global declaration (i.e. can not have two versions of it at the same time), where as it is really a module. the change should be done on DefinitelyTyped to address this.

similar discussion can be found under https://github.com/Microsoft/TypeScript/issues/12550

**Comment by mhegazy** [comment]
I have a fix up in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/13382

**Comment by mhegazy** [comment]
should be fixed by DefinitelyTyped/DefinitelyTyped#13382
